### PR TITLE
Development

### PR DESCRIPTION
**Pull request by gferrarocamus**
Associations Project: Private Events

### CODE REVIEW

File: app/views/users/show.html.erb [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+<% if !@user.created_events.empty? %>
+  <h3>My events:</h3>
+  <ul>
+  <% @user.created_events.each do |event| %>
+    <li><%= link_to event.name, event_path(event.id) %></li>
+  <% end %>
+  </ul>
+<% end %>
+
+<% if !@previous_events.empty? %>
```
Comment (gferrarocamus) [inline app/views/users/show.html.erb]: #2

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    if @user.save
+      log_in @user
+      flash[:info] = 'Welcome!'
+      redirect_to profile_url
+    else
+      render 'new'
+    end
+  end
+
+  def show
```
Comment (gferrarocamus) [inline app/controllers/users_controller.rb:23]: #2

File: app/models/event.rb [hunk 3]
```diff
@@ -0,0 +1,23 @@
+# frozen_string_literal: true
+
+# Event
+class Event < ApplicationRecord
+  has_many :attendance, foreign_key: :attended_event_id, inverse_of: :attendances, dependent: :destroy
```
Comment (heyapricot) [inline app/models/event.rb]: Can you explain further the idea behind using `inverse_of`?

**Review by heyapricot — COMMENTED** [review]
Current status (May 2nd):
- [ ] The project has a descriptive README.md 
### Authentication
- [x] The application has a functional `login` page (It doesn't require authentication according to the Odin Project's description)
- [x] The application uses either a cookie or a session to log in the user.
### Navigation
- [ ] The application includes navigation links on top
### Users

#### Model
- [ ] It has a method `previous_events` to retrieve past events
- [ ] It has a method `upcoming_events` to retrieve upcoming events

#### Actions
##### `new`
- [x] It has a view with a form that creates a new User instance when submitted
##### `show` 
- [ ] Lists all the events the user has created
- [ ] Lists all the events where the `user` is an attendee
- [ ] The events are separated in past events and upcoming events. 

### Events
#### Model
- [x] It has a class method to retrieve all past events (`Event.past`)
- [x] It has a class method to retrieve all upcoming events (`Event.upcoming`)
- [x] The class methods use scopes
#### Actions
##### `new`
- [x] It has a view with a form that creates a new Event instance when submitted. The creator of the event is `current_user`
##### `index`
- [x] Lists all events separated into Past and Upcoming categories.  

##### `show` 
- [ ] It displays the creator of the event
- [ ] It displays the list of attendees

File: app/controllers/events_controller.rb [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    @past_events = Event.past
+    @upcoming_events = Event.upcoming
+  end
+
+  def show
+    @event = Event.find(params[:id])
+  end
+
+  def new
+    @user = current_user
```
Comment (heyapricot) [inline app/controllers/events_controller.rb]: Do you need `@user` here?

File: app/views/users/show.html.erb [hunk 5]
```diff
@@ -0,0 +1,29 @@
+<h2><%= @user.name %></h2>
+
+<% if !@user.created_events.empty? %>
+  <h3>My events:</h3>
+  <ul>
+  <% @user.created_events.each do |event| %>
```
Comment (heyapricot) [inline app/views/users/show.html.erb]: Your view is making two calls to `@user.created_events`. You can store the result of that query on an instance variable in the controller. Make your view use as less logic as you can

File: app/views/events/show.html.erb [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+<p>Location: <%= @event.location %></p>
+<p><%= @event.description unless @event.description.nil? %></p>
+<br>
+
+<%= link_to "I'm attending!", attendance_path(@event.id), method: :post unless (current_user.nil? || user_attending) %>
+
+<% if !@event.event_attendees.empty? %>
+  <h3>Attendees:</h3>
+  <ul>
+  <% @event.event_attendees.each do |attendee| %>
```
Comment (heyapricot) [inline app/views/events/show.html.erb]: Store `event_attendees` in an instance variable in the controller so you avoid calling it twice from the view

**Review by heyapricot — CHANGES_REQUESTED** [review]
# It's not a two-way road 🛣
👋🏼 Hello team! There is an issue on your app related to the use of `inverse_of` that is not allowing me to assign attendees to an event. 
You also are putting a lot of logic in the views. Please fix that. 
I've raised some issues on the repo and added comments. I've also commented on the current status of the project so you can have an overview of the things that are missing. Please complete the missing stuff, address the comments and after that submit a new code review request. Thanks!

File: app/models/event.rb (hunk 3 above)
Comment (gferrarocamus) [reply app/models/event.rb]: We included inverse_of because of the linter Rubocop which was requiring it. We'll revise this!

**Comment by gferrarocamus** [comment]
Thanks for the feedback, @heyapricot! We've made the required changes and submitted a new review request.

File: app/models/user.rb [hunk 7]
```diff
@@ -0,0 +1,19 @@
+# frozen_string_literal: true
+
+# User
+class User < ApplicationRecord
+  has_many :attendances, class_name: 'Attendance',
```
Comment (rvvergara) [inline app/models/user.rb]: Why did we need to put `class_name: 'Attendance`` here? The table name is already :attendances. We only have to put the class_name if the association is called by a different name.

File: app/models/event.rb [hunk 8]
```diff
@@ -0,0 +1,24 @@
+# frozen_string_literal: true
+
+# Event
+class Event < ApplicationRecord
+  has_many :attendances, class_name: 'Attendance',
```
Comment (rvvergara) [inline app/models/event.rb]: No need to put `class_name: 'Attendance'`

File: app/controllers/sessions_controller.rb [hunk 9]
```diff
@@ -0,0 +1,24 @@
+# frozen_string_literal: true
+
+# SessionsController
+class SessionsController < ApplicationController
+  def new; end
+
+  def create
```
Comment (rvvergara) [inline app/controllers/sessions_controller.rb:7]: You need error handling here supposing makes a mistake signing in or if a user tries to press login without placing a username

**Review by rvvergara — CHANGES_REQUESTED** [review]
Hi @gferrarocamus and @jeton-th ,

Receiving the baton from @heyapricot. From someone who first saw your Private Events app here are some thoughts (on top of the comments I already placed in the respective code):
1. You really need to work on your README. You can use templates such as [this](https://github.com/RichardLitt/standard-readme)
2. As a signed in user I don't see the events created by other users. 
3. As an event creator, how do I add attendees to the event? 
4. When I for example log on with a wrong name (non-registered name), I get the following error:
image

**Comment by gferrarocamus** [comment]
Hi, @rvvergara! Thank you for your feedback. We've updated with project following your indications. 

As [the project specifications](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations) don't mention how attendees should be added to events, we decided to use a button that allows each user to select what events they are attending (instead of having creators add attendees).

image

We've submitted a new review request now. Thanks!

**Review by rvvergara — CHANGES_REQUESTED** [review]
Hi @gferrarocamus and @jeton-th,

Awesome. I'm not getting any errors anymore when I click 'Create my account' without providing any name. With regards to attending events, my bad, I didn't realize the 'Im attending' button was there. I was looking for the event creator to add the attendees. Your implementation is good.

Though there are just 3 minor changes I request this time:

1. Show a flash message when I try to create an account without a name. The flash message should tell me what I did wrong. Right now it's just re-rendering the Signup page without any information
2. Please see below image:

image

The upcoming event which is closer should appear first. As it is right now (in the image) the event that will happen in July is shown ahead of the event happening this May 24th. 

Also it would be nice if you could format the dates into a more readable format (say, May 24, 2019 instead of 2019-05-24). You can use [this](https://apidock.com/ruby/DateTime/strftime) 

3. Yes still about your README. Currently it's still the default provided by Rails. Please edit it accordingly. I already provided you in my previous review a template you can follow

**Comment by gferrarocamus** [comment]
Thanks, @rvvergara! We'll submit a new review request after adding the flash (1) and fixing the dates (2). As for (3), you can see the updated readme [here](https://github.com/gferrarocamus/private-events/blob/development/README.md).

**Review by rvvergara — CHANGES_REQUESTED** [review]
Hi,

I think we must have missed something. When im accessing the app as a new user (or if the app has no data yet) the following error happens:

image

Since you are telling your routes to redirect you to views/index.html.erb and that file is currently dependent on the existence of events data then this error is happening. 

Another thing I should have already mentioned to you guys is (now since I suddenly had to look into your views/events/index.html.erb file) is that instead of running 2 `each` statements for your past events and upcoming events you should have created an `event` partial which you will just render in lieu of the each statements. Please refer to Section 3.4.8 of the [Rails Guides](https://guides.rubyonrails.org/v5.2/layouts_and_rendering.html) on how you can do this. This will make your code much cleaner.

**Review by rvvergara — APPROVED** [review]
hi @gferrarocamus and @jeton-th,

It's all good now! Great job.. I like how you did the `views/events/index.html.erb` rendering of events in conjunction with your events_controller. 

Also the bug I previously mentioned is extinct. It's all working like a well oiled machine now.

You may now merge this into `master` and submit your project completion.

# :+1:

**Pull request merged**